src/routes/transactions/helpers/imitation-transactions.helper.ts

```diff
@@ -0,0 +1,104 @@
+import { IConfigurationService } from '@/config/configuration.service.interface';
+import { TransactionItem } from '@/routes/transactions/entities/transaction-item.entity';
+import {
+  isTransferTransactionInfo,
+  TransferDirection,
+} from '@/routes/transactions/entities/transfer-transaction-info.entity';
+import { isErc20Transfer } from '@/routes/transactions/entities/transfers/erc20-transfer.entity';
+import { Inject } from '@nestjs/common';
+
+export class ImitationTransactionsHelper {
+  private readonly vanityAddressChars: number;
+
+  constructor(
+    @Inject(IConfigurationService) configurationService: IConfigurationService,
+  ) {
+    this.vanityAddressChars = configurationService.getOrThrow(
+      'mappings.imitationTransactions.vanityAddressChars',
+    );
+  }
+
+  /**
+   * Filters out outgoing ERC20 transfers that imitate their direct predecessor in
+   * value and vanity (but not the exact) address of the recipient.
+   *
+   * @param transactions - list of transactions to filter
+   * @param previousTransaction - transaction to compare last {@link transactions} against
+   *
+   * Note: this only handles singular imitation transfers. It does not handle multiple
+   * imitation transfers in a row, nor does it compare batched multiSend transactions
+   * as the "distance" between those batched and their imitation may not be immediate.
+   */
+  filterOutgoingErc20ImitationTransfers(
+    transactions: Array<TransactionItem>,
+    previousTransaction: TransactionItem | undefined,
+  ): Array<TransactionItem> {
+    // TODO: Instead of filtering, mark transactions so client can display them differently
+    return transactions.filter((item, i, arr) => {
+      // Executed by Safe - cannot be imitation
+      if (item.transaction.executionInfo) {
+        return item;
+      }
+
+      const prevItem = i === arr.length - 1 ? previousTransaction : arr[i + 1];
+
+      // No reference transaction to filter against
+      if (!prevItem) {
+        return item;
+      }
+
+      const txInfo = item.transaction.txInfo;
+      const prevTxInfo = prevItem.transaction.txInfo;
+
+      if (
+        // Only consider transfers...
+        !isTransferTransactionInfo(txInfo) ||
+        !isTransferTransactionInfo(prevTxInfo) ||
+        // ...of ERC20s...
+        !isErc20Transfer(txInfo.transferInfo) ||
+        !isErc20Transfer(prevTxInfo.transferInfo)
+      ) {
+        return item;
+      }
+
+      // ...that are outgoing
```

Comment:
The pattern I've seen was 1) a legit outgoing transfer 2) incoming scam transfers.

<img width="989" alt="Screenshot 2024-05-08 at 10 43 17" src="https://github.com/safe-global/safe-client-gateway/assets/381895/82d3c852-6973-4411-b9cd-637a3e9a9e5e">

Reply:
As incoming transfers are legitimate, [we proposed](https://www.notion.so/safe-global/Address-poisoning-heuristics-01153b329f544e9cb92384ebf87f70d5?pvs=4#e3171dc299bd4cc39d57a283ae104052) only filtering/marking outgoing ones.

Reply:
Not sure I understand why all the complicated logic is needed for just outgoing txs then. Aren't they trivially detectable by a missing/bad nonce?

Reply:
It doesn't matter about the type. Any transaction that transfers an ERC-20 [will emit a `Transfer` event](https://eips.ethereum.org/EIPS/eip-20#transfer-1), which is the "reference marker" used by imitators.